ipa-core/src/protocol/basics/validate.rs

```diff
@@ -0,0 +1,218 @@
+};
+
+type HashFunction = Sha256;
+type HashSize = <HashFunction as OutputSizeUser>::OutputSize;
+type HashOutputArray = [u8; <HashSize as Unsigned>::USIZE];
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+struct HashValue(GenericArray<u8, HashSize>);
+
+impl Serializable for HashValue {
+    type Size = HashSize;
+    type DeserializationError = Infallible;
+
+    fn serialize(&self, buf: &mut GenericArray<u8, Self::Size>) {
+        buf.copy_from_slice(self.0.as_slice())
+    }
+
+    fn deserialize(buf: &GenericArray<u8, Self::Size>) -> Result<Self, Self::DeserializationError> {
+        Ok(Self(buf.to_owned()))
+    }
+}
+
+impl Message for HashValue {}
+
+struct ReplicatedValidatorFinalization<C> {
+    f: Pin<Box<dyn Future<Output = Result<(), Error>>>>,
+    ctx: C,
+}
+
+impl<C: Context + 'static> ReplicatedValidatorFinalization<C> {
+    fn new(active: ReplicatedValidatorActive<C>) -> Self {
+        let ReplicatedValidatorActive {
+            ctx,
+            left_hash,
+            right_hash,
+        } = active;
+        // Ugh: The version of sha2 we currently use doesn't use the same GenericArray version as we do.
+        let left_hash = HashValue(GenericArray::from(<HashOutputArray>::from(
+            left_hash.finalize_fixed(),
+        )));
+        let right_hash = HashValue(GenericArray::from(<HashOutputArray>::from(
+            right_hash.finalize_fixed(),
+        )));
+        let left_peer = ctx.role().peer(Direction::Left);
+        let right_peer = ctx.role().peer(Direction::Left);
+        let ctx_ref = &ctx;
+
+        let f = Box::pin(async move {
+            try_join(
+                ctx_ref
+                    .send_channel(left_peer)
+                    .send(RecordId::FIRST, left_hash.clone()),
+                ctx_ref
+                    .send_channel(right_peer)
+                    .send(RecordId::FIRST, right_hash.clone()),
+            )
+            .await?;
+            let (left_recvd, right_recvd) = try_join(
+                ctx_ref.recv_channel(left_peer).receive(RecordId::FIRST),
+                ctx_ref.recv_channel(right_peer).receive(RecordId::FIRST),
+            )
+            .await?;
+            if left_hash == left_recvd && right_hash == right_recvd {
+                Ok(())
+            } else {
+                Err(Error::Internal) // TODO add a code
+            }
+        });
+        Self { f, ctx }
+    }
+
+    fn poll(&mut self, cx: &mut TaskContext<'_>) -> Poll<Result<(), Error>> {
+        self.f.poll_unpin(cx)
+    }
+}
+
+struct ReplicatedValidatorActive<C> {
+    ctx: C,
+    left_hash: Sha256,
+    right_hash: Sha256,
+}
+
+impl<C: Context + 'static> ReplicatedValidatorActive<C> {
+    fn new(ctx: C) -> Self {
+        Self {
+            ctx,
+            left_hash: HashFunction::new(),
+            right_hash: HashFunction::new(),
+        }
+    }
+
+    fn update<S, V>(&mut self, s: &S)
+    where
+        S: ReplicatedSecretSharing<V>,
```

Comment:
not sure if it is worth it, but you may just require `V: Serializable` here rather than `SharedValue`

Reply:
`ReplicatedSecretSharing` requires that `V: SharedValue` so that doesn't really do anything.